### PR TITLE
Add LiveKit audio transport

### PR DESCRIPTION
**Pull request by joachimchauvet**
This might not be perfect yet but I would love to get some initial feedback.
It looks like there are quite a few people interested #325

### CODE REVIEW

File: src/pipecat/transports/services/livekit.py
Comment (aconchillo) [inline src/pipecat/transports/services/livekit.py:29]: We probably want to add `livekit` to pyproject.toml with the right dependencies.

**Comment by aconchillo** [comment]
This looks really good! I'll take a final look later, thank you!!!

**Comment by cyrilS-dev** [comment]
This looks amazing @joachimchauvet ! I'm particularly interested in this PR as I'm using LiveKit. 
I will give it a try as soon as possible and provide feedback. Looking forward to testing it out!
Many thanks

**Comment by joachimchauvet** [comment]
> @joachimchauvet I just tested and I can't play the audio when I join the room using the LK playground. I join the same room as my agent, I have both participants active in the room, the room events are consistent, but the 'Agent connected' status on the playground remains pending, as does the audio track.

It seems to work fine on my side. 
However, I noticed that I forgot to adjust the sample rate in https://github.com/joachimchauvet/pipecat-livekit/blob/main/examples/foundational/01b-livekit-audio.py
Could it be that the voice was playing so fast that you did not really hear it if you were using that example?
Are you using a different token for the bot and the user?

**Comment by cyrilS-dev** [comment]
@joachimchauvet i'm using a different token for the bot and the user yes

**Comment by cyrilS-dev** [comment]
@joachimchauvet can you please check : 
https://github.com/cyrilS-dev/pipecat/commit/0a64aba7addde9dd6f113eacd11e7fb83468d585
https://github.com/cyrilS-dev/pipecat/commit/89260736ae9e84e98f85f5abae56572bbd6dbb9c

This version uses LiveKit Agents worker and job logic.

**Comment by aconchillo** [comment]
This looks great! Please, rebase and make sure the pipeline is green. We are now using Ruff as our formatter. Make sure everything is green and I think we cna merge after that. Thank you! :pray:

**Comment by cyrilS-dev** [comment]
@joachimchauvet I encountered an error while using the LiveKitInputTransport class in my application :

`Error in audio input task: 'LiveKitInputTransport' object has no attribute '_internal_push_frame'`

`_internal_push_frame` method is not defined in the LiveKitInputTransport class

**Comment by cyrilS-dev** [comment]
> > @joachimchauvet I encountered an error while using the LiveKitInputTransport in my application :
> > `Error in audio input task: 'LiveKitInputTransport' object has no attribute '_internal_push_frame'`
> > `_internal_push_frame` method is not defined in the LiveKitInputTransport class
> 
> Were you doing something special or made changes to the `LiveKitInputTransport`? It inherits the `BaseInputTransport` so `_internal_push_frame` should be defined. The method works fine on my side 🤔

There is no `_internal_push_frame` method into `BaseInputTransport` since [https://github.com/pipecat-ai/pipecat/commit/337f048864d154dc3a70e4ae38eb3fc54f69abbe#diff-75a719845c4cd87dda07b88df06ba32275ebc7be1fdcbbd21fa87554d3a800af](https://github.com/pipecat-ai/pipecat/commit/337f048864d154dc3a70e4ae38eb3fc54f69abbe#diff-75a719845c4cd87dda07b88df06ba32275ebc7be1fdcbbd21fa87554d3a800af)

**Comment by cyrilS-dev** [comment]
@joachimchauvet I encountered an error in the `_audio_in_task_handler` method of the `LiveKitInputTransport` class., using Deepgram as STT. :
```
AttributeError: 'AsyncListenWebSocketClient' object has no attribute '_socket'
2024-09-25 15:49:15.863 | ERROR    | pipecat.processors.frame_processor:push_frame:203 - Uncaught exception in LiveKitInputTransport#0: 'AsyncListenWebSocketClient' object has no attribute '_socket'
```
The issue arises when  `await self.push_frame(pipecat_audio_frame)` attempts to push frames before the Deepgram WebSocket connection is fully initialized.

**Comment by aconchillo** [comment]
> @joachimchauvet I encountered an error in the `_audio_in_task_handler` method of the `LiveKitInputTransport` class., using Deepgram as STT. :
> 
> ```
> AttributeError: 'AsyncListenWebSocketClient' object has no attribute '_socket'
> 2024-09-25 15:49:15.863 | ERROR    | pipecat.processors.frame_processor:push_frame:203 - Uncaught exception in LiveKitInputTransport#0: 'AsyncListenWebSocketClient' object has no attribute '_socket'
> ```
> 
> The issue arises when `await self.push_frame(pipecat_audio_frame)` attempts to push frames before the Deepgram WebSocket connection is fully initialized.

In theory, all processors should receive the `StartFrame` first (this is guaranteed now in `main`) and they should make sure any connection or whatever is needed is established so it's safe to push frames after.

It's possible that this branch hasn't been rebased so it doesn't have that `StartFrame` guarantee. Would it be possible for you to rebase locally and try that? I could be wrong and maybe it's something else...

**Comment by cyrilS-dev** [comment]
Everything is running perfectly; just one detail: `LiveKitInputTransport` stops when the room is disconnected, but `LiveKitOutputTransport` doesn't; this triggers an error after a while from the TTS service: `no close frame received or sent`.

**Review by aconchillo — APPROVED** [review]

**Comment by aconchillo** [comment]
> Everything is running perfectly; just one detail: `LiveKitInputTransport` stops when the room is disconnected, but `LiveKitOutputTransport` doesn't; this triggers an error after a while from the TTS service: `no close frame received or sent`.

I'll wait on @joachimchauvet to answer this before merging.

**Comment by aconchillo** [comment]
I'll go ahead and merge this one. If there's any error you can provide a fix later. Thank you!!!

**Comment by joachimchauvet** [comment]
> Everything is running perfectly; just one detail: `LiveKitInputTransport` stops when the room is disconnected, but `LiveKitOutputTransport` doesn't; this triggers an error after a while from the TTS service: `no close frame received or sent`.

I don't have this error with my agents. I have an idea where that might come from but I was not able to reproduce that `no close frame received or sent` directly. @cyrilS-dev could you share a pipeline that triggers this error?

**Comment by cyrilS-dev** [comment]
@joachimchauvet The execution is blocked when calling `await super().stop(frame)` in the `stop `method of the `BaseOutputTransport` class.

The `stop `method in the `BaseOutputTransport `class is causing the execution to hang due to : 

```python
if self._sink_task:
     await self._sink_task
if self._sink_clock_task:
     await self._sink_clock_task
```

As a result, the `await self._client.disconnect()` call is never reached.

@aconchillo do you have any insights on this? Thanks

**Pull request merged**
